app/resolve-guid/route.ts

```diff
@@ -1,17 +1,102 @@
+import { action, computed } from '@ember-decorators/object';
 import { service } from '@ember-decorators/service';
 import Route from '@ember/routing/route';
+import Transition from '@ember/routing/transition';
+import { camelize } from '@ember/string';
+import Ember from 'ember';
 import DS from 'ember-data';
+import Features from 'ember-feature-flags/services/features';
+import config from 'ember-get-config';
+import param from 'ember-osf-web/utils/param';
+
+interface RouterMicroLib {
+    generate(routeName: string, ...args: any[]): string;
+}
+
+interface PrivateRouter {
+    _routerMicrolib: RouterMicroLib;
+    _engineInfoByRoute?: Record<string, any>;
+    location: Ember.EmberLocation;
+}
+
+const GUID_REGEX = /^[23456789abcdefghjkmnpqrstuvwxyz]{5}$/;
+
+const { featureFlagNames: { routes } } = config;
 
 export default class ResolveGuid extends Route {
+    @service features!: Features;
     @service store!: DS.Store;
+    _router!: PrivateRouter; // tslint:disable-line:variable-name
+
+    @computed(`features.${camelize(routes['registries.overview'])}`)
+    get routeMap(): Record<string, string> {
+        return {
+            file: 'guid-file',
+            node: 'guid-node',
+            preprint: 'guid-preprint',
+            user: 'guid-user',
+            registration: this.features.isEnabled(routes['registries.overview'])
+                ? 'registries.overview'
+                : 'guid-registration',
+        };
+    }
+
+    isEngineRoute(route: string): boolean {
+        return Boolean(this._router._engineInfoByRoute && route in this._router._engineInfoByRoute);
+    }
+
+    looksLikeGUID(guid: string): boolean {
+        return GUID_REGEX.test(guid);
+    }
+
+    generateURL(route: string, ...args: any[]): string {
+        // NOTE: The router's urlFor is skipped over here as it passes the result of generate into the location
+        // implementation, which would rip out the "--PATH" which is used for routing here.
+        return this._router._routerMicrolib.generate(route, ...args);
+    }
+
+    async beforeModel(transition: Transition) {
+        const params = Object.assign(
+            { guid: '', path: '' },
+            ...Object.values(transition.params),
+        );
 
-    async model(this: ResolveGuid, params: { guid: string }): Promise<void> {
-        // Block until the Guid is resolved, then always transition somewhere else.
-        try {
+        if (!this.isEngineRoute(params.guid) && !this.looksLikeGUID(params.guid)) {
+            return this.replaceWith('not-found');
+        }
+
+        let expanded: string;
+
+        // If guid is an engine route, transition into it
+        if (this.isEngineRoute(params.guid)) {
```

Comment:
This confused me at first, but it helped to realize this special routing with `--` segments is being used for two distinct situations:
- guid routing
- mounting engines at `/` (actually at `/--my-engine/`, but effectively `/`)

The latter case feels brittle. It matches the "guid" with the *name* of the route, not its path. This works for registries but could easily break if we tried to use it more generally.